### PR TITLE
[dynamo] Support multiple inheritance for custom dict construction

### LINKED ISSUE #141118: Dynamo: how to deal with multiple inheritance (nn.Module/MutableMapping)?
### 🐛 Describe the bug

TensorDict is a MutableMapping object, and is treated as such by torch.compile:
```python
import torch
from tensordict import TensorDict

td = TensorDict(a=1, b=2, c=True)

@torch.compile(fullgraph=True)
def add1(td):
    return TensorDict(**td)+1

add1(td)
```

We also have a `TensorDictParams` primitive that acts a bit like ParameterList: it is a TensorDict but also an nn.Module. That's useful when you want to set a TensorDict in an nn.Module have have the leaf tensors included in the state_dict, or dispatch ops like `module.to(...)` to the tensors it contains. However, `_dynamo` looks at it like an nn.Module and not a MutableMapping
```python
import torch
from tensordict import TensorDictParams, TensorDict

td = TensorDictParams(TensorDict(a=1, b=2, c=True))

@torch.compile(fullgraph=True)
def add1(td):
    return TensorDict(**td)+1

add1(td)
```
breaks with
```
  File "/Users/vmoens/venv/rl/lib/python3.10/site-packages/torch/_dynamo/variables/dicts.py", line 357, in call_method
    dict_vt = BuiltinVariable.call_custom_dict(tx, dict, args[0])
  File "/Users/vmoens/venv/rl/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 1432, in call_custom_dict
    unimplemented(f"{user_cls.__name__}(): {args} {kwargs}")
  File "/Users/vmoens/venv/rl/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 313, in unimplemented
    raise Unsupported(msg, case_name=case_name)
torch._dynamo.exc.Unsupported: dict(): (UnspecializedNNModuleVariable(TensorDictParams),) {}
```

My understanding is that `call_custom_dict` looks at the arg an in one case it's a `variables.MutableMappingVariable` which is fine but in the other it's a `UnspecializedNNModuleVariable` which isn't a mutable mapping.

So I guess my question is (other than how can we fix this) how does dynamo look at multiple inheritance? Shouldn't there be a way to tell "look, this isn't a bird or a fish but a fish that can fly"? 

(note that in this specific case, `smth(**obj)` will call `obj.keys()` followed by `obj.__getitem__` which are ops that compile is happy about - maybe that's what `call_custom_dict` should be doing?)

Here is a MRE:
```python
import torch
from torch import nn
import collections

# class MyWeirdDict(collections.abc.MutableMapping):  # Works
class MyWeirdDict(collections.abc.MutableMapping, nn.Module):  # breaks
    def __init__(self, **kwargs):
        super().__init__()
        self._items = kwargs
    def keys(self):
        return self._items.keys()
    def __getitem__(self, item):
        return self._items[item]
    def __setitem__(self, key, value):
        self._items[key] = value
    def __delitem__(self, item):
        del self._items[item]
    def __len__(self):
        return len(self._items)
    def __iter__(self):
        yield from self._items
    def __hash__(self):
        return hash(id(self))
    def items(self):
        for k, v in self._items.items():
            yield (k, v)

@torch.compile(fullgraph=True)
def to_weird_dict(td):
    return MyWeirdDict(**td)

d = MyWeirdDict(a=1, b=2, c=3)
to_weird_dict(d)
```

### Error logs

See above

### Versions

nightlies

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142416

This patch applies a local and practical workaround for custom dict
construction when multiple inheritance is involved.

Handling multiple inheritance in general could be a lot more involved,
so I created #142414 to track that.

Fixes #141118.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by StrongerXi** [comment]
Newly added test fails internally because `inline_inbuilt_nn_modules` is False there. I added some updates to address that.

File: torch/_dynamo/variables/dicts.py [hunk 1]
```diff
@@ -344,33 +337,25 @@ def call_method(
...
-                    TupleVariable,
-                    ListIteratorVariable,
-                    variables.IteratorVariable,
-                    UserDefinedObjectVariable,
-                ),
-            )
-
-            is_kwargs_supported = len(kwargs) > 0 and len(args) == 0
-
-            if is_args_supported or is_kwargs_supported:
```
Comment (StrongerXi) [inline torch/_dynamo/variables/dicts.py:361]: `BuiltinVariable.call_custom_dict` takes care of all these instance checks and fail with `unimplemented` for unsupported cases, so we don't need to duplicate the check here. Checking here also restricted this from hitting the new `elif hasattr(arg, "value") and isinstance(arg.value, MutableMapping):` branch for more general custom dict support.
 
https://github.com/pytorch/pytorch/blob/604c5a9e29eb9939d3b3538fd5b04753d7aeeef5/torch/_dynamo/variables/builtin.py#L1373-L1397

File: torch/_dynamo/variables/nn_module.py [hunk 2]
```diff
@@ -131,18 +131,18 @@ class NNModuleVariable(VariableTracker):
...
 
     def __init__(
-        self, module_type: type, module_key: str, module: torch.nn.Module, **kwargs
+        self, module_type: type, module_key: str, value: torch.nn.Module, **kwargs
     ) -> None:
         super().__init__(**kwargs)
         self.module_type = module_type
         self.module_key = module_key
-        self.module = module
+        self.value = value
```
Comment (StrongerXi) [inline torch/_dynamo/variables/nn_module.py:145]: Simple rename to make this compatible with the new `elif hasattr(arg, "value") and isinstance(arg.value, MutableMapping):` branch in `BuiltinVariable.call_custom_dict`.

**Comment by StrongerXi** [comment]
Update expected result for `add_loop_inductor`, `-1.5%` instruction count, so it's improvements.

**Comment by StrongerXi** [comment]
Reverting the benchmark change after chatting with @laithsakka offline.

**Pull request closed without merging**